### PR TITLE
Fix checking symbolic shapes inside torch._check

### LINKED ISSUE #110719: Make torch._check work in Dynamo
### 🐛 Describe the bug

as stated

### Versions

main

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### PR DESCRIPTION
**Pull request by isuruf**
Fixes https://github.com/pytorch/pytorch/issues/110719#issuecomment-1768710678

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -159,13 +159,13 @@ def is_concrete_bool(a: Union[bool, SymBool]):
...
 
-def _iterate_exprs(val: Union[SymInt, torch.Tensor]) -> Iterable[sympy.Expr]:
+def _iterate_exprs(val: Union[SymInt, torch.Tensor]) -> Iterable[sympy.Basic]:
     if isinstance(val, SymTypes):
         # This allow applies to the jagged layout NestedTensor case as
         # singleton ints are not symbolic
         if is_symbolic(val):
             yield val.node.expr
-    elif isinstance(val, sympy.Expr):
+    elif isinstance(val, sympy.Basic):
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:168]: Not sure what's going on here

**Review by ezyang — APPROVED** [review]
Wow! That was much easier than I expected

File: torch/_inductor/codegen/common.py [hunk 2]
```diff
@@ -314,6 +314,9 @@ def _print_Pow(self, expr):
     def _print_Unequality(self, expr):
         return " != ".join(map(self.paren, map(self._print, expr.args)))
 
+    def _print_Equality(self, expr):
+        return " == ".join(map(self.paren, map(self._print, expr.args)))
```
[the remark this answers is not in the document]
Comment (isuruf) [reply torch/_inductor/codegen/common.py]: Thanks. I updated the PR and added a test that checked all relational operators.

**Review by peterbell10 — APPROVED** [review]

**Comment by jeanschmidt** [comment]
Please check https://github.com/pytorch/pytorch/actions/runs/6899362830/job/18771441277

**Comment by ezyang** [comment]
The False is True error likely root causes to https://github.com/pytorch/pytorch/issues/113905

So maybe we should do the Eq workaround @asmeurer suggested

**Comment by isuruf** [comment]
I can't reproduce that failure locally nor on CI here.

**Comment by isuruf** [comment]
Any ideas on how to reproduce the failure?

**Comment by malfet** [comment]
@isuruf you should be to download docker on your machine to repro the failure. Or feel free to ping me on Slack

**Comment by isuruf** [comment]
Running `PYTORCH_TEST_WITH_INDUCTOR=1 python test/test_torch.py -k test_lazy_clone_cuda_bool`, I can reproduce the failure even on main (c435b8c10a1b9652a35e1a87cee6c3942356660d) with this PR reverted.

**Comment by malfet** [comment]
> Running `PYTORCH_TEST_WITH_INDUCTOR=1 python test/test_torch.py -k test_lazy_clone_cuda_bool`, I can reproduce the failure even on main ([c435b8c](https://github.com/pytorch/pytorch/commit/c435b8c10a1b9652a35e1a87cee6c3942356660d)) with this PR reverted.

I wonder if this is a compiler artifact caching (see https://github.com/pytorch/pytorch/issues/110098 ) @zou3519 have you landed the changes that would invalidate the cache?

**Comment by malfet** [comment]
I think the offending test should be excluded: https://github.com/pytorch/pytorch/pull/114012

**Pull request closed without merging**